sdk/python/feast/feature_service.py

```diff
@@ -86,29 +86,48 @@ def __init__(
                 self.feature_view_projections.append(feature_grouping.projection)
 
     def infer_features(self, fvs_to_update: Optional[Dict[str, FeatureView]] = None):
+        """
+        Infers the features for the projections of this feature service, and updates this feature
+        service in place.
+
+        This method is necessary since feature services may rely on feature views which require
+        feature inference.
+        """
         for feature_grouping in self._features:
             if isinstance(feature_grouping, BaseFeatureView):
-                # For feature services that depend on an unspecified feature view, apply inferred schema
+                projection = feature_grouping.projection
+
                 if fvs_to_update and feature_grouping.name in fvs_to_update:
-                    if feature_grouping.projection.desired_features:
-                        desired_features = set(
-                            feature_grouping.projection.desired_features
-                        )
+                    # There are three situations to be handled. First, the projection specifies
+                    # desired features, in which case we should select those desired features.
+                    # Second, the projection does not specify any desired features but has
+                    # already selected features, in which case nothing needs to be done. And
+                    # third, the projection does not specify any desired features but has not 
+                    # yet selected features (since the original feature view did not yet have
+                    # features), in which case we should select all possible inferred features.
+                    if projection.desired_features:
+                        # First case, so we select the specific desired features.
+                        desired_features = set(projection.desired_features)
                         actual_features = set(
                             [
                                 f.name
                                 for f in fvs_to_update[feature_grouping.name].features
                             ]
                         )
                         assert desired_features.issubset(actual_features)
+
                         # We need to set the features for the projection at this point so we ensure we're starting with
                         # an empty list.
-                        feature_grouping.projection.features = []
+                        projection.features = []
                         for f in fvs_to_update[feature_grouping.name].features:
                             if f.name in desired_features:
-                                feature_grouping.projection.features.append(f)
+                                projection.features.append(f)
+                    elif not projection.desired_features and projection.features:
```

Comment:
would be cleaner imo if you had early returns instead of the elif and else.. and maybe some example definitions

e.g. 
```python
if projection.desired_features:
  # The projection wants to reference inferred features. Validate they exist
   # Example: FeatureService(features=[[fv_with_no_schema["feature]])
   ...
   return

if projection.features:
  # The projection only references features from a FV's known schema (not inferred). 
  # Example 1: FeatureService(features=[fv_with_schema[["feature"]])
  # Example 2: FeatureService(features=[fv_with_schema])
  return

# The projection wants all features from a FV that has an inferred schema
# Example: FeatureService(features=[fv_with_no_schema])
...
```